Addressing feedback after UMT (#881)
o1-labs/docs2: docs/berkeley-upgrade/mainnet-database-maintenance.mdx

O1Labs and Mina Foundation teams have always made every effort to ensure that the software they produce are thoroughly tested and of the highest quality. 
However, we are only humans and we make mistakes. Such mistake occurred during work on the Mina version that was used on the mainnet for some time. 
We fixed the errors quickly. However, in decentralized world  archive nodes are not very forgiving and may contain historical issues. We are talking exclusively about 2 issues:

- Missing / Invalid Nonces - a historical issue which skewed nonces in balances table. Issue was resolved but you may still have nonces that are missing or invalid
- Incorrect ledger hashes - a historical issue which had the same root cause as 'Missing / Invalid Nonces' one. However, the outcome is that a 'replayer run' (which a operation of validating archive node against deamon ledger) is showing ledger mismatches and cannot pass problematic blocks unless ran with flag `--continue-on-error` (which can hide other issues) 

On top of that, we never forget about one problem that occurs constantly and is the nightmare of all archive node operators

> - Missing blocks - everlasting issue which is cause by a disturbance in deamon node operations which can cause that archive node won't receive all blocks from network thath is supposed to hold. As a result chain continuity in archive database can be broken.

Reviewer: ```suggestion
- Missing blocks - everlasting issue which is cause by a disturbance in deamon node operations which can cause that archive node won't receive all blocks from network that it is supposed to hold. As a result chain continuity in archive database can be broken.
```

---
title: Devnet/Mainnet database maintenance
sidebar_label: Devnet/Mainnet database maintenance
hide_title: true
description: Steps to properly maintain correctness of archive database
keywords:
  - Berkeley
  - upgrade
  - archive migration
  - planning 
  - prerequisites
  - mina archive node
  - archive node
  - mainnet
  - devnet
  - database
---

# Devnet/Mainnet database in context of Berkeley Upgrade

After the berkeley migration, orginal devnet/mainnet database is not needed anymore unless you are interested in 
preserving some aspect of the database which will be lost during the migration process.

After the successful migration you will have two databases:

> - orginal devnet/mainnet database - with small data adjustments ( all pending blocks from last canoncial block till fork block will be converted to canoncial )

Reviewer: ```suggestion
- Original Devnet/Mainnet database - with small data adjustments (all pending blocks from last canonical block till fork block will be converted to canonical)
```

> - new berkeley database based on devnet/mainnet data but:

Reviewer: ```suggestion
- New Berkeley database based on Devnet/Mainnet data but:
```

>   - without devnet/mainnet orphaned blocks

Reviewer: ```suggestion
  - without Devnet/Mainnet orphaned blocks
```

>   - without devnet/mainnet pending blocks which are not a part of canoncial chain

Reviewer: ```suggestion
  - without Devnet/Mainnet pending blocks which are not a part of canonical chain
```

>   - with all pending blocks on canoncial chain converted to canoncial block

Reviewer: ```suggestion
  - with all pending blocks on canonical chain converted to canonical block
```

##### Missing blocks 

The daemon node unavailability can cause the archive node to miss some of the blocks. 

If you are uploading the missing blocks to Google Cloud, the missing blocks can be reapplied from precomputed blocks and preserve chain continuity. 

To automatically verify and patch missing blocks, use the [download_missing_blocks.sh](https://raw.githubusercontent.com/MinaProtocol/mina/2.0.0berkeley_rc1/src/app/rosetta/download-missing-blocks.sh) script. Because the `download-missing-blocks` script uses localhost as the database host, unless you modify `PG_CONN` in `download_missing_block.sh`, the script assumes that psql is running on localhost on port 5432.

1. Install the required `mina-archive-blocks` and `mina-missing-blocks-auditor` scripts that are packed in the `minaprotocol/mina-archive:1.4.0-c980ba8-bullseye` Docker image.

2. Export the `BLOCKS_BUCKET`:

   ```sh
   export BLOCKS_BUCKET="https://storage.googleapis.com/my_bucket_with_precomputed_blocks"

3. Run the `mina-missing-blocks-auditor` script from the database host:

   For devnet:
   ```sh

>    download-missing-blocks.sh devnet {db_user} {db_password}

Reviewer: should use packaged & not source name, right?

If you uploaded the missing blocks to Google Cloud, the missing blocks can be reapplied from precomputed blocks to preserve chain continuity. 

1. To automatically verify and patch missing blocks, use the [download_missing_blocks.sh](https://raw.githubusercontent.com/MinaProtocol/mina/2.0.0berkeley_rc1/src/app/rosetta/download-missing-blocks.sh) script. 

   The `download-missing-blocks` script uses `localhost` as the database host so the script assumes that psql is running on localhost on port 5432. Modify `PG_CONN` in `download_missing_block.sh` for your environment.

1. Install the required `mina-archive-blocks` and `mina-missing-blocks-auditor` scripts that are packed in the `minaprotocol/mina-archive:1.4.0-c980ba8-bullseye` Docker image.

1. Export the `BLOCKS_BUCKET`:

    ```sh
    export BLOCKS_BUCKET="https://storage.googleapis.com/my_bucket_with_precomputed_blocks"
    ```

1. Run the `mina-missing-blocks-auditor` script from the database host:
    
    For Devnet:
    
    ```sh
    download-missing-blocks.sh devnet {db_user} {db_password}
    ```

    For Mainnet:
    
    ```sh
    download-missing-blocks.sh mainnet {db_user} {db_password}
    ```
    

> ## What steps are required for the Berkeley migration?

Reviewer 2: the title of this heading does not seem to match the paragraph below... what are we trying to say here?

Reviewer 3: It's trying to say you need to make sure there's no missing blocks.
